crates/rpc/src/pending.rs

```diff
@@ -11,40 +20,234 @@ use tokio::sync::watch::Receiver as WatchReceiver;
-    pub state_update: Arc<StateUpdate>,
-    pub number: BlockNumber,
+    block: Arc<PendingBlockVariant>,
+    state_update: Arc<StateUpdate>,
+    number: BlockNumber,
 }
 
 impl PendingData {
+    pub fn from_pending_block(
+        block: PendingBlock,
+        state_update: StateUpdate,
+        number: BlockNumber,
+    ) -> Self {
+        Self {
+            block: Arc::new(PendingBlockVariant::Pending(block)),
+            state_update: Arc::new(state_update),
+            number,
+        }
+    }
+
+    /// Converts a pre-confirmed block fetched from the gateway into pending
+    /// data.
+    ///
+    /// Candidate transactions are filtered out and handled separately. State
+    /// update is constructed from the per-transaction updates.
+    pub fn from_pre_confirmed_block(
+        mut block: starknet_gateway_types::reply::PreConfirmedBlock,
+        number: BlockNumber,
+    ) -> Self {
+        // Get rid of Nones in transaction receipt
+        let transaction_receipts: Vec<_> =
+            block.transaction_receipts.into_iter().flatten().collect();
+        let number_of_pre_confirmed_transactions = transaction_receipts.len();
+        let candidate_transactions = block
+            .transactions
+            .split_off(number_of_pre_confirmed_transactions);
+
+        // Compute aggregated state diff.
+        let mut state_diff = starknet_gateway_types::reply::state_update::StateDiff::default();
+        for transaction_diff in block.transaction_state_diffs.into_iter().flatten() {
+            state_diff.extend(transaction_diff);
+        }
+        state_diff.deduplicate();
+        let state_update = starknet_gateway_types::reply::StateUpdate {
+            state_diff,
+            block_hash: Default::default(),
+            new_root: StateCommitment::default(),
+            old_root: StateCommitment::default(),
+        };
+        let state_update = StateUpdate::from(state_update);
+
+        let block = PreConfirmedBlock {
+            l1_gas_price: block.l1_gas_price,
+            l1_data_gas_price: block.l1_data_gas_price,
+            l2_gas_price: block.l2_gas_price,
+            sequencer_address: block.sequencer_address,
+            status: Status::PreConfirmed,
+            timestamp: block.timestamp,
+            starknet_version: block.starknet_version,
+            l1_da_mode: block.l1_da_mode.into(),
+            transactions: block.transactions,
+            transaction_receipts,
+        };
+
+        Self {
+            block: Arc::new(PendingBlockVariant::PreConfirmed(
+                block,
+                candidate_transactions,
+            )),
+            state_update: Arc::new(state_update),
+            number,
+        }
+    }
+
+    pub fn block_number(&self) -> BlockNumber {
+        self.number
+    }
+
+    /// Returns a mutable reference to the block number.
+    ///
+    /// To be used in tests only.
+    pub fn block_number_mut(&mut self) -> &mut BlockNumber {
```

Comment:
`#[cfg(test)]` should work here - right?

Reply:
Indeed, fixed. (For some reason I thought the test was in a different crate -- but that's not the case.)